### ISSUE TITLE
data_grid raises an error when extra_dims has more than one entry

### ISSUE THREAD

**Issue by briandleahy** [manoharan-lab/holopy#254]
Calling data_grid when ``extra_dims`` contains more than one element can sometimes raise an exception. The issues seems to arise because ``holopy.core.metadata.data_grid`` and ``holopy.core.metadata.make_coords`` assume that iteration over Python dictionaries occurs in the insertion order, which is not the case for pre-Python 3.6 versions. 

Here is a minimum working example, which raises an error on my machine:

```python
import numpy as np
from holopy.core.metadata import data_grid

ar = np.empty((1, 1, 1, 1, 2, 3, 4, 5, 6, 7, 8))
extra_dims = {
    name: np.arange(size) for name, size in zip('abcdefgh', ar.shape[3:])}

grid = data_grid(ar, spacing=1, z=0, extra_dims=extra_dims)
# ValueError: conflicting sizes for dimension 'g': length 1 on the data but length 7 on coordinate 'g'
```

Replacing the ``dict`` with an ``OrderedDict`` causes the tests to pass, 
```python
from collections import OrderedDict

import numpy as np
from holopy.core.metadata import data_grid

ar = np.empty((1, 1, 1, 1, 2, 3, 4, 5, 6, 7, 8))

extra_dims = OrderedDict(
    [[name, np.arange(size)] for name, size in zip('abcdefgh', ar.shape[3:])])

grid = data_grid(ar, spacing=1, z=0, extra_dims=extra_dims)
```
... but I'm not sure that this fixes the problem because it seems that ``holopy.core.metadata.data_grid`` and ``make_coords`` internally cast an OrderedDict to a dict anyways.

Incidentally I noticed this because some of the new tests for the 3.3 branch were failing on my machine and passing on @barkls 's.

**Comment by briandleahy** [comment]
After PR #287 I ran the tests on my Python 3.5 machine and several of them failed for similar reasons on ordered vs unordered dicts. For instance, here are 2 failing tests:
```python
    bg = BoundedGaussian(*parameters.values())
  File "/home/brian/holopy/holopy/inference/prior.py", line 255, in __init__
    "Lower bound {} must be less than mean {}. Upper bound {} must"
holopy.scattering.errors.ParameterSpecificationError: Lower bound {} must be less than mean {}. Upper bound {} must be greater than mean.
```
and 
```python
    u = Uniform(*parameters.values())
  File "/home/brian/holopy/holopy/inference/prior.py", line 103, in __init__
    lower_bound, upper_bound))
holopy.scattering.errors.ParameterSpecificationError: Lower bound 2 is not less than upper bound 1
```
both of which fail because they unpack dicts and expect the order to be preserved. So I think rather than trying to swim upstream with this, I'm thinking that we just be thankful that issue #254 didn't cause problems when people were running pre-3.6, and we support holopy only on python 3.6+. I think we can state this in either the README.rst or the documentation.

Thoughts?

**Comment by briandleahy** [comment]
OK, pursuant to our conversations, I've cast everything to an OrderedDict (on my local branch, which I will submit a PR for soon). However, this problem can still be recreated by calling ```data_grid``` with an unordered dictionary (as above). In this case, the casting as an OrderedDict doesn't guarantee that the iteration order of ```extra_dims``` will be the same as the dimension order of ```ar```. So it's not solvable within data_grid; it's only solvable in the methods that call it, and right now it will be solved in the methods that call ```data_grid```.

I think the solution is to leave it, make data_grid a "private" class method not part of the API (which it's not), and to refactor data_grid later into a factory class method where what it is called with is more strongly controlled.

**Comment by vnmanoharan** [comment]
Independent of this solution, I would suggest that we require python 3.7+ for HoloPy version 3.3 (3.7+ because dictionary insertion order is *guaranteed* to be preserved in 3.7 and above, but not 3.6+).  Rationale: we don't have the bandwidth to support multiple python versions, and with conda being cross-platform and multi-environment, it's not a huge burden on the user to install 3.7 if needed.

**Comment by briandleahy** [comment]
Agreed. I don't think we want to waste time maintaining functionality across multiple versions.

**Comment by briandleahy** [comment]
Fixed / rendered irrelevant by #291
